crates/eip4844/src/kzg_open.rs

```diff
@@ -0,0 +1,177 @@
+use bls12_381::{ff::Field, Scalar};
+use polynomial::poly_coeff::PolyCoeff;
+
+pub(crate) fn bitreverse(mut n: u32, l: u32) -> u32 {
+    let mut r = 0;
+    for _ in 0..l {
+        r = (r << 1) | (n & 1);
+        n >>= 1;
+    }
+    r
+}
+
+pub(crate) fn bitreverse_slice<T>(a: &mut [T]) {
+    if a.is_empty() {
+        return;
+    }
+
+    let n = a.len();
+    let log_n = n.ilog2();
+    assert_eq!(n, 1 << log_n);
+
+    for k in 0..n {
+        let rk = bitreverse(k as u32, log_n) as usize;
+        if k < rk {
+            a.swap(rk, k);
+        }
+    }
+}
+
+/// Compute evaluation and quotient by Ruffini's rule.
+pub fn compute_evaluation_and_quotient(polynomial: &PolyCoeff, z: Scalar) -> (Scalar, Vec<Scalar>) {
+    let mut quotient = vec![Scalar::ZERO; polynomial.len()];
+    (0..polynomial.len() - 1)
+        .rev()
+        .for_each(|i| quotient[i] = quotient[i + 1] * z + polynomial[i + 1]);
+    let y = quotient[0] * z + polynomial[0];
```

Comment:
Can we pull this out into a function so that we can test it in isolation, like:


```rust
/// Divides poly by X-Z using ruffini's rule
fn divide_by_linear(poly: &[Scalar], z: Scalar) -> Vec<Scalar> {
    let mut quotient: Vec<Scalar> = Vec::with_capacity(poly.len());
    let mut k = Scalar::from(0u64);

    for coeff in poly.iter().rev() {
        let t = *coeff + k;
        quotient.push(t);
        k = z * t;
    }

    // Pop off the remainder term
    quotient.pop();

    // Reverse the results as monomial form stores coefficients starting with lowest degree
    quotient.reverse();
    quotient
}
```

It adds more lines of code, but makes it easier to read and reason about